### PR TITLE
Added lots of decoder tests and fixed lots of bugs.

### PR DESCRIPTION
**Pull request by haberman**

### CODE REVIEW

**Comment by haberman** [comment]
To @scwhittle. I'm not going to worry about the Coveralls "failure."

File: tests/pb/test_decoder.cc [hunk 1]
```diff
@@ -558,9 +571,9 @@ uint32_t Hash(const string& proto, const string* expected_output, size_t seam1,
 
 void CheckBytesParsed(const upb::pb::Decoder& decoder, size_t ofs) {
   // We could have parsed as many as 10 bytes fewer than what the decoder
-  // previously accepted, since we can buffer up to 10 partial bytes internally
+  // previously accepted, since we can buffer up to 12 partial bytes internally
```
Comment (scwhittle) [inline tests/pb/test_decoder.cc]: Lots of different numbers 10, 12, 14 
should they all be 14?  Maybe with explanation on where 14 was calculated

Comment (haberman) [inline tests/pb/test_decoder.cc]: Done.

File: upb/pb/decoder.c [hunk 2]
```diff
@@ -257,15 +286,14 @@ size_t upb_pbdecoder_suspend(upb_pbdecoder *d) {
     d->ptr = d->residual;
     return 0;
   } else {
-    size_t consumed;
+    size_t ret = d->size_param - (d->end - d->checkpoint);
```
Comment (scwhittle) [inline upb/pb/decoder.c]: size_t unconsumed

Comment (scwhittle) [inline upb/pb/decoder.c]: size_t consumed

File: upb/pb/decoder.int.h [hunk 3]
```diff
@@ -220,9 +220,9 @@ struct upb_pbdecoder {
 
   /* Buffer for residual bytes not parsed from the previous buffer.
    * The maximum number of residual bytes we require is 12; a five-byte
```
Comment (scwhittle) [inline upb/pb/decoder.int.h]: 14

Comment (haberman) [inline upb/pb/decoder.int.h]: Done.

**Pull request merged**